rfcs/control-groups/README.md

```diff
@@ -0,0 +1,279 @@
+settings and is compared with number of entries in the `ProcessList`.
+`max_concurrent_queries_for_user` is implemented with a different data structure
+called `ProcessListForUser`. [3] contains additional examples.
+
+
+## Design
+
+There are a lot of expectations from "resource pools", and different customers
+have different needs. The current goal is not to address all of them but rather
+lay a foundation that would allow implementing basic means for resource isolation
+today and not preclude future extensions with things like statistical multiplexing
+of CPU/IO/Network, preemptions based on query priorities and so on.
+
+
+### What we want
+
+We want to address the problems mentioned above. Provide flexibility for configuring
+limits for a single query, for all queries of a user, for queries of a group of users
+and have implementation consistency.
+
+First thing that comes to mind in this space is [Wiki: Linux Control Groups](https://en.wikipedia.org/wiki/Cgroups)
+([kernel.org: cgroup-v2.txt](https://www.kernel.org/doc/Documentation/cgroup-v2.txt)).
+
+> cgroup is a mechanism to organize processes hierarchically and distribute
+> system resources along the hierarchy in a controlled and configurable manner.
+> cgroup is largely composed of two parts - the core and controllers.
+>
+> cgroup core is primarily responsible for hierarchically organizing processes.
+>
+> A cgroup controller is usually responsible for distributing a specific type of
+> system resource along the hierarchy although there are utility controllers
+> which serve purposes other than resource distribution.
+>
+> cgroups form a tree structure and every process in the system belongs to
+> one and only one cgroup.
+
+![cgroupv2 hierarchy](./cgroupv2-hierarchy.jpg)
+
+
+### What we have
+
+ClickHouse already has things similar to cgroup controllers that are used to
+implement various limits.
+
+Examples: [MemoryTracker][MemoryTracker.h], Network [Throttler][Throttler.h].
+
+[ProfileEvents][ProfileEvents.h] and thread groups hierarchies with
+[ThreadProfileEvents][ThreadProfileEvents.h] could fall in this category as well
+even though today only tracking is supported (rather than "controlling").
+[3] has some neat ideas on how it could be possible to "bind" actions to events.
+
+An important property to mention of the above mentioned "things" is that all of
+them are hierarchical.
+
+The missing part is that there isn't a lot of flexibility in configuring these.
+
+
+### A rough sketch of how the future could look like
+
+Let's draw some inspiration from Linux control groups and adapt the existing
+"resource controllers" to achieve "resource pools".
+
+* There should be a root control group for the system limits
+
+* It should be possible to define control groups for background processes
+and assign tables to these control groups.
+    * Ie a table or a group of tables could use at most `X b/s` of disk, 
+    `Y b/s` of network for replication, `Z b` of storage.
```

Comment:
>There is immediate tradeoff between speed and isolation if we want to add some guarantees for certain groups. We would need to leave some resources reserved for this groups to alliow them to get realtime result if they send there query in next moment.

Not necessarily. Queries can be "paused" to free some cpu/disk, while memory can indeed be reserved (usually reserving memory should not be a problem, unlike CPU/disk).

About pausing, right now each query executor is a bunch of threads and if we are talking about non-voluntary preempting they can just receive SIGSTOP, but yes this may expires some timeouts or some time/speed-related settings and once the query will be continued it will "instantly" fails.

Another option is to add voluntary preempting, but this will require adding yielding in more places.